Add initial implementation

willdurand opened the pull request:
This first version reports information to Telemetry for eTLDs (public suffixes) and `webRequest` as per the PRD. Once loaded, the extension should do its thing automatically. It listens to search engine modifications and updates itself accordingly.

Some more information about this patch (in no particular order):

- We only record an event when `from` and `to` values are different (the values are eTLDs fwiw) to comply with the data review
- When we find a redirect caused by an extension (_via_ `webRequest`), we record an event and we don't check for server-side redirects
- We follow redirect chains and only record an event when the first and last URLs have different eTLDs
- We only record an event when we have an add-on ID (or more than one)
- In case of multiple extensions, we record an event for each add-on ID

willdurand committed:
add very first version

mixedpuppy commented on extension/api.js:
```diff
@@ -0,0 +1,154 @@
+        Services.obs.addObserver(
+          onSearchEngineModifiedObserver,
+          SearchUtils.TOPIC_ENGINE_MODIFIED
+        );
```
You don't need an object here, just use:
```
let observer = (aSubject, aTopic, aData) {}
Services.obs.addObserver(
          observer,
          SearchUtils.TOPIC_ENGINE_MODIFIED
        );
```

Then you can remove the startup/shutdown and tracking the object

mixedpuppy commented on extension/api.js:
```diff
@@ -0,0 +1,154 @@
+        async getAddonById(addonId) {
+          const addon = await AddonManager.getAddonByID(addonId);
+
+          return { version: addon.version };
```
```
if (addon) {
  return { version: addon.version };
}
```

no need to return null

mixedpuppy commented on extension/api.js:
```diff
@@ -0,0 +1,154 @@
+        // of the given URL.
+        // See: https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIEffectiveTLDService
+        async getPublicSuffix(url) {
+          try {
+            return Services.eTLD.getBaseDomain(Services.io.newURI(url));
+          } catch (err) {
+            Cu.reportError(err);
+            return null;
+          }
```
I suppose this is fine, but since it's async, you could .catch() it in the background script

willdurand replied:
I tried to keep this API "simple" to use for the background script.

mixedpuppy replied:
returning a promise is the standard mechanism for extension APIs, so promise.catch would also be the standard approach.  Again, not a big deal.

mixedpuppy commented on extension/api.js:
```diff
@@ -0,0 +1,154 @@
+);
+
+this.addonsSearchExperiment = class extends ExtensionAPI {
+  onStartup() {}
```
drop this

willdurand replied:
mm it's there because otherwise `ExtensionCommon` says:

> api.onStartup is not a function

willdurand replied:
nevermind, @rpl found the cause... (`startup` event was registered in `manifest.json`)

mixedpuppy commented on extension/api.js:
```diff
@@ -0,0 +1,154 @@
+  onStartup() {}
+
+  makeOnSearchEngineModified(context) {
+    return new ExtensionCommon.EventManager({
```
not a big deal, but I'd just put this in the api itself

```
onSearchEngineModified:  new ExtensionCommon.EventManager({ ... }),
```

mixedpuppy commented on extension/api.js:
```diff
@@ -0,0 +1,154 @@
+        // `getAddonById()` returns add-on details if it exists. Note that it
+        // does not return a full `add-on` object but a minimal object with
+        // only the necessary information.
+        async getAddonById(addonId) {
```
This is only returning the addon version.  Maybe call this getAddonVersion, and rather than return an object just return the version.

mixedpuppy commented on extension/background.js:
```diff
@@ -0,0 +1,230 @@
+    // This one is needed in addition to `onBeforeRequest` because this
+    // extension might be registered before or after some extensions.
+    // Depending on that, our `onBeforeRequest` listener might not be called,
+    // which is why we also listen to `onBeforeRedirect`.
```
This would also be used to see the server side redirects

mixedpuppy commented on extension/background.js:
```diff
@@ -0,0 +1,230 @@
+  // `redirectUrl` is usually valid when the redirect has been detected via
+  // `onBeforeRedirect`, otherwise it's likely `undefined`.
+  webRequestHandler = async ({ requestId, url, redirectUrl }) => {
+    if (this.lastRequestIdReported === requestId) {
+      console.debug(`request ID '${requestId}' already reported, skipping...`);
+      return;
+    }
```
You might get interleaved calls so this isn't the best way to protect against double reporting.  I'm thinking that we do not need the listener for onBeforeRequest.  If something redirects in that call, you should still get a call to onBeforeRedirect.

mixedpuppy commented on extension/background.js:
```diff
@@ -0,0 +1,230 @@
+    if (isServerSideRedirect) {
+      const id = this.getAddonIdFromUrl(url);
+
+      // We shouldn't report built-in search engines.
+      if (!id.endsWith("@search.mozilla.org")) {
```
Why?  Aside from that, non of our builtin engines have code so this isn't necessary.

willdurand replied:
In a previous iteration, there was no check between `from` and `to` eTLDs so the extension was reporting the `google.com` search engine in some cases.

mixedpuppy commented on extension/background.js:
```diff
@@ -0,0 +1,230 @@
+      (await browser.addonsSearchExperiment.getRequestUrl(requestId));
+    const to = await browser.addonsSearchExperiment.getPublicSuffix(requestUrl);
+
+    if (from === to) {
+      // We do not report redirects to same public suffixes.
+      return;
+    }
```
I'm not certain that we shouldn't log this.

willdurand replied:
I don't think we're interested in extensions redirecting from `example.com` to `example.com` or `foo.example.com` (ETLD still being `example.com`) to `example.com` because we "stay" on the same domain.

cc @wagnerand

wagnerand replied:
At least for the first iteration, we want to log only when the domain is different. Non-domain changes in the url need more discussion and aligning before we can add logging for them.

mixedpuppy replied:
I feel like if some other extension is changing the request we should note that.  So it's not about a redirect from a.com to a.com, but a redirect from a.com to a.com caused by some random extension.  I understand that isn't apparent to users, but to me it does indicate an extension messing with a search request.

mixedpuppy commented on extension/background.js:
```diff
@@ -0,0 +1,230 @@
+
+const start = async () => {
+  const exp = new AddonsSearchExperiment();
+  await exp.monitor();
```
I don't think it's necessary to await here (or in the later call), so the `start` function could also be dropped.

willdurand replied:
mm what if the listener below is called before the first `monitor()` call ends?

mixedpuppy commented on extension/api.js:
```diff
@@ -0,0 +1,154 @@
+          },
+        };
+
+        Services.obs.addObserver(
```
Might be good to wait for search initialization before adding the observer.  On firefox startup, that will prevent churn when the service is adding engines.

https://searchfox.org/mozilla-central/rev/9c9f2f85d00aef1943cb521ac95c72bae932ebc5/browser/components/extensions/parent/ext-browser.js#289

willdurand committed:
address most comments

willdurand commented:
@mixedpuppy so I addressed most of the comments (thanks!) and left a few comments where it made sense to reply **but** it seems the `wrapper` cannot be retrieved anymore with only a `onBeforeRedirect` listener.

Now that I think about it, I remember having a similar problem when I just started playing with privileged add-ons and I mentioned it to @rpl I think. I thought it was just a problem with `web-ext` or something but now, with the same setup, I don't get any registered channel for a `requestId` when there is only a `onBeforeRedirect` listener. It does not make much sense to write such a thing because this does not make a lot of sense either but that's what I am observing now 🤷‍♂️

willdurand committed:
remove log statement

willdurand commented:
> Now that I think about it, I remember having a similar problem when I just started playing with privileged add-ons and I mentioned it to @rpl I think. I thought it was just a problem with `web-ext` or something but now, with the same setup, I don't get any registered channel for a `requestId` when there is only a `onBeforeRedirect` listener. It does not make much sense to write such a thing because this does not make a lot of sense either but that's what I am observing now 🤷‍♂️

Actually, looking at the code again... A while ago, @rpl found that we had to call `registerTraceableChannel()` to retrieve a channel wrapper later (in this extension for example). This does not seem to be the case when we do NOT register a listener with a `blocking` or `policy` option: https://searchfox.org/mozilla-central/source/toolkit/components/extensions/webrequest/WebRequest.jsm#876-888

`onBeforeRedirect` cannot be blocking so I guess that's why it'd work when a block `onBeforeRequest` listener is added? It's just a somewhat educated guess, I don't really know what's going on in `WebRequest.jsm`.

willdurand committed:
Listen to onBeforeRequest with a no-op handler

willdurand commented:
> `onBeforeRedirect` cannot be blocking so I guess that's why it'd work when a block `onBeforeRequest` listener is added? It's just a somewhat educated guess, I don't really know what's going on in `WebRequest.jsm`.

Registering a no-op listener for `onBeforeRequest` seems to work as intended... Not sure if that's what we want, though.

willdurand committed:
remove onStartup()

willdurand commented:
Requesting a review from @Standard8 to make sure that the use of the "Search" APIs are correct here.

mixedpuppy commented:
> > `onBeforeRedirect` cannot be blocking so I guess that's why it'd work when a block `onBeforeRequest` listener is added? It's just a somewhat educated guess, I don't really know what's going on in `WebRequest.jsm`.
> 
> Registering a no-op listener for `onBeforeRequest` seems to work as intended... Not sure if that's what we want, though.

RIght, I forgot about that.  I think a no-op listener is fine for onBeforeRequest, and having the actual logic in onBeforeRedirect.  While it's unfortunate to have to do that, it does simplify things.

mixedpuppy commented on extension/background.js:
```diff
@@ -0,0 +1,195 @@
+    return this.matchPatternsMap;
+  }
+
+  async monitor() {
```
Add a comment explaining this function.  Somethign along the lines "If the search service changes the set of engines that are enabled, we update our pattern matching in the webrequest listeners"

willdurand committed:
add comment

willdurand commented:
Requesting a general review from @Rob--W

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,149 @@
+XPCOMUtils.defineLazyGlobalGetters(this, ["ChannelWrapper"]);
+
+XPCOMUtils.defineLazyServiceGetter(
+  Services,
```
Define on `this`, not on `Services`.

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,149 @@
+const { SearchUtils } = ChromeUtils.import(
+  "resource://gre/modules/SearchUtils.jsm"
+);
+const { WebRequest } = ChromeUtils.import(
```
This import is currently unused; but see a later comment (you may need it anyway).

willdurand replied:
(forgot to remove that, indeed)

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,149 @@
+/* global ExtensionCommon, ExtensionAPI, Services, XPCOMUtils */
+const { SearchUtils } = ChromeUtils.import(
```
This is used only once to import one string. You could also just declare a constant with the value to minimize dependencies.

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,149 @@
+              if (aTopic !== SearchUtils.TOPIC_ENGINE_MODIFIED) {
+                return;
+              }
+
```
The background script is only interested in `engine-added` / `engine-removed`, let's filter it here and let the background unconditionally invoke the `monitor` method.

willdurand replied:
Sure, that works too.

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,198 @@
+    // an add-on ID corresponding to the add-on that initiated the redirect.
+    // It might not return anything when the redirect is a search server-side
+    // redirect but it can also be caused by an error.
+    let addonId = await browser.addonsSearchExperiment.getRequestProperty(
```
Doing this from a `webRequest.onBeforeRedirect` handler is not reliable, because the event is non-blocking.

To get reliable results, you would have to register a listener from the `api.js` script and retrieve the `ChannelWrapper` from there, synchronously (and get the property of interest).

You can look at ext-webRequest.js for inspiration; ignore the part about primed listeners (that logic is only relevant for extensions that need to intercept startup requests):

- Registering / unregistering internal WebRequest listeners: https://searchfox.org/mozilla-central/rev/5227b2bd674d49c0eba365a709d3fb341534f361/toolkit/components/extensions/parent/ext-webRequest.js#89-94
- Implementing an event listener: https://firefox-source-docs.mozilla.org/toolkit/components/extensions/webextensions/events.html
- Other examples of `WebRequest.jsm` consumers in unit tests: https://searchfox.org/mozilla-central/search?q=WebRequest.on&path=extensions%2Ftest%2F&case=true&regexp=false

willdurand replied:
done

Rob--W reviewed (commented):
See also pm

willdurand committed:
address comments

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,148 @@
+            };
+
+            searchInitialized.then(() => {
+              Services.obs.addObserver(
```
`addObserver` should be called synchronously, to make sure that it balances the `removeObserver call later.

If you want to exclude search engines from before the initialization, check `Services.search.isInitialized` before calling `fire.async()`.

willdurand replied:
@Rob--W something like this?

```diff
diff --git a/extension/api.js b/extension/api.js
index 4886d1f..f030c8c 100644
--- a/extension/api.js
+++ b/extension/api.js
@@ -124,15 +124,15 @@ this.addonsSearchExperiment = class extends ExtensionAPI {
                 return;
               }
 
-              fire.async();
+              searchInitialized.then(() => {
+                fire.async();
+              });
             };
 
-            searchInitialized.then(() => {
-              Services.obs.addObserver(
-                onSearchEngineModifiedObserver,
-                SEARCH_TOPIC_ENGINE_MODIFIED
-              );
-            });
+            Services.obs.addObserver(
+              onSearchEngineModifiedObserver,
+              SEARCH_TOPIC_ENGINE_MODIFIED
+            );
 
             return () => {
               Services.obs.removeObserver(
```

willdurand committed:
move searchInitialized in the listener

willdurand committed:
register onBeforeRedirect handler in the privileged API to be sure that we can read the channel properties

willdurand committed:
Add comment

willdurand committed:
add a comment about public suffixes belonging to same entity

willdurand commented:
I think I addressed all the comments so far. I added some more comments for some of the things that we decided not to take into account right now.

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,189 @@
+                return;
+              }
+
+              searchInitialized.then(() => {
```
I thought that your goal was to skip initial search engine modifications? The logic here wouldn't skip them, but delay the event dispatch.

Check for the `Services.search.isInitialized` if you want to check a boolean value in order to skip if needed.

willdurand replied:
The goal was to make sure the search service is initialized before we can create and use an event in the background script. We don't necessarily need to skip any logic. I guess `searchInitialized` might not even be needed here because it is also used in `getMatchPatterns()`.

willdurand committed:
fix listener deinit

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,189 @@
+            // See: toolkit/components/extensions/parent/ext-webRequest.js
+            let filter2 = {};
+            if (filter.urls) {
+              let perms = new MatchPatternSet([
```
perms not used, let's remove it.

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,189 @@
+
+            // See: toolkit/components/extensions/parent/ext-webRequest.js
+            let filter2 = {};
+            if (filter.urls) {
```
In the webRequest API this option is not optional. Let's make it unconditional.

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,189 @@
+
+            WebRequest.onBeforeRedirect.addListener(
+              listener,
+              filter2,
```
The filter should also include `types: ["main_frame"]` since you're only interested in top-level navigation requests.

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,189 @@
+              {
+                addonId: extension.id,
+                policy: extension.policy,
+                blockingAllowed: extension.hasPermission("webRequestBlocking"),
```
Make this unconditionally false. The event doesn't support blocking listeners anyway.

willdurand committed:
more changes

willdurand committed:
Add support for multiple extension IDs for a same URL pattern

willdurand commented:
It isn't impossible that two search engines register similar URL templates to the point where the URL patterns would be the same. For that reason, we now have a map where the key is the URL pattern and the value is a list of IDs (as opposed to just an ID before).

willdurand commented:
There is one caveat right now: we don't handle redirect chains at the moment, only the first redirect is taken into account. I am looking into it.

Standard8 commented on extension/api.js:
```diff
@@ -0,0 +1,193 @@
+
+          try {
+            await searchInitialized;
+            const visibleEngines = await Services.search.getVisibleEngines();
```
Per discussions, I think it is better to use `getEngines()` here. Might want to rename the variable as well.

Standard8 reviewed (approved):
The use of search APIs here looks fine to me, with the one modification mentioned.

willdurand committed:
Add support for redirect chains

willdurand committed:
Use getEngines()

willdurand committed:
60 seconds

willdurand commented:
Rob said the one timer and its delay might be a bit fragile/too low so I bumped the delay from 10s to 60s. It might still be a bit fragile because if the redirect chain takes more than 60s (slow connections?), we wouldn't necessarily catch a server redirect. OTOH, we could probably increase this delay if needed. He suggested a different approach where we'd register a listener for every "hop" in the redirect chain. I am going to look into that but I have the feeling that the single listener approach could be a good start anyway.

willdurand committed:
refactor the logic to follow redirect chains

willdurand committed:
refactor listeners

willdurand commented:
>  He suggested a different approach where we'd register a listener for every "hop" in the redirect chain. I am going to look into that but I have the feeling that the single listener approach could be a good start anyway.

We discussed this idea over Slack and after a few iterations, we went with the following approach: the first request of interest will register a `onBeforeRequest` wildcard listener that will follow all subsequent requests of interest as well as the first one. This wildcard listener and the old request IDs to follow are cleaned up later, either via a `onCompleted` wildcard listener or a fallback function with a timer.

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,281 @@
+    // This is needed to force the registration of a traceable channel.
+    browser.webRequest.onBeforeRequest.addListener(
+      this.onRequestHandler,
+      { urls: patterns },
```
`types: ["main_frame"],`

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,281 @@
+        types: ["main_frame"],
+        urls: ["<all_urls>"],
+      });
+    }
```
The `requestId` and timer should be registered here. Otherwise, the wildcard listeners will be around indefinitely.

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,281 @@
+      delete this.requestIdsToFollow[requestId];
+    }
+
+    if (Object.keys(this.requestIdsToFollow).length === 0) {
```
Nit: If `requestIdsToFollow` is a `Map` instead of a plain object, then you can simply use `.size` instead of `Object.keys(...).length === 0`.

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,281 @@
+    this.onRedirectHandler({
+      requestId,
+      url: redirectUrl,
+      redirectUrl: url,
```
During `onBeforeRequest`, `this.requestIdsToFollow[requestId].redirectUrl` is the destination URL at the time of the `onBefireRedirect` event. That would be the same URL as `url` here.

Originally, my idea was that you would use blocking `onBeforeRequest` to track all redirects, like this example:

1. onBeforeRequest - match pattern matched, add wildcard listener
2. onBeforeRequest 1..n - observed 1..n redirection targets
3. onCompleted or timed out: remove wildcard listener.

Only the first `onBeforeRequest` is actually necessary. So you could simplify the implementation by doing it like this:

1. onBeforeRequest - match pattern matched - register `this.requestIdsToFollow`. Register wildcard `onBeforeRedirect` (note: ...Redirect, not ...Request).
2. onBeforeRedirect - if `this.requestIdsToFollow` is set, process the request.
3. onCompleted or timed out: remove wildcard listener.

Rob--W reviewed (commented):
+followup in pm

willdurand committed:
simplify how we handle redirect chains a bit

willdurand committed:
Update listeners when an engine has changed (useful for local dev)

willdurand committed:
send 1 event per redirect chain with the initial/last eTLDs

willdurand committed:
cleanups

willdurand committed:
some more cleanups/comments

willdurand committed:
set debug mode when extension is loaded temporarily

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,195 @@
+                Cu.reportError(err);
+              }
+
+              fire.async({ addonId, redirectUrl, requestId, url });
```
Make this `fire.sync`, so that the event is not unnecessarily delayed, which could potentially result in out-of-order events relative to the `fire.sync` call from the `webRequest.onBeforeRequest` event.

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,290 @@
+      this.debug(`start following requestId=${requestId}`);
+
+      // Pass metadata to the follow listener.
+      this.requestIdsToFollow.set(requestId, {
```
This would unconditionally overwrite an existing chain. See pm.

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,290 @@
+  }
+}
+
+browser.runtime.onInstalled.addListener(({ temporary: debugMode }) => {
```
The `onInstalled` event doesn't fire once the add-on has been installed. Consequently the add-on wouldn't work any more. The `AddonSearchExperiment` construction should be unconditional.

willdurand replied:
ha, shoot. Well in this case, I am going to remove that and I'll toggle the boolean manually...

willdurand committed:
more fixes

willdurand committed:
only include search urls in the list of urls to monitor

willdurand committed:
do not report built-in search engines

willdurand committed:
fix comments

Rob--W reviewed (approved):
Outstanding remarks have been addressed or acknowledged and the current initial implementation lgtm.

willdurand committed:
remove await

willdurand committed:
fix comment

willdurand commented:
We plan to address the following points in new PRs:

1. we will report a false positive when (1) an extension updates the query params of a request like google.com (which shouldn't be reported) and then (2) the user uses the "I'm lucky" feature (which does a server-side redirect). This happens because (1) will set the extension ID in the properties bag of the request.
2. we should probably restrict the monitoring to requests from the location bar, bookmarks and other system requests. This is possible by checking `originUrl` in the listeners.

willdurand commented:
@mixedpuppy assuming it looks good to you, can we merge this patch and add the necessary config/files to build the extension with TaskCluster, etc. so that we can ask QA to verify it? (and also because it would be easier to iterate with smaller patches in the future)

willdurand committed:
typo

willdurand committed:
Only execute onRedirectHandler if requestId isn't being followed yet

willdurand committed:
remove useless async

willdurand committed:
remove under useless occurrence of async

willdurand committed:
simplify implementation with a stop listener on the channel

willdurand commented:
@Rob--W after talking to @mixedpuppy who told me about [`loadInfo` and `redirectChain`](https://searchfox.org/mozilla-central/source/toolkit/components/url-classifier/nsUrlClassifierUtils.cpp#682-685), we decided to use that instead of manually following a request and building the redirect chain ourselves. The first problem was to find how to use that in the background script (where the rest of the logic lives) and it turns out that we can listen to `stop` event on a channel, which is exactly what we need. These two things allow us to expose a new event named `onRedirected`, fired at the end of a request, and the event will pass all the needed data to the listener. This simplifies the background code a lot and that makes things easier to follow IMO. WDYT?

willdurand committed:
update readme

willdurand committed:
use good old bind instead of class property for onRedirectedListener

willdurand commented:
(rebased this PR to get CI, fixed a linter issue)

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,233 @@
+
+              fire.sync({
+                addonId,
+                firstUrl: redirectChain && redirectChain[0],
```
`firstUrl` is not necessarily the matched URL, it could be a non-matching URL that redirects to a URL that is matched by the `onBeforeRedirect` listener. Consequently, the logic in `background.js` will not be able to find the addonId associated with the request.

You should probably do the mapping of the matched URL to addonIds in the `onBeforeRedirect` listener instead of the background script (or at least pass the matched URL to the event listener).

willdurand replied:
heh correct, so we probably don't care about the first url of the chain, we only care about the first matched url.

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,233 @@
+                context.extension.policy,
+                context.xulBrowser.frameLoader.remoteTab
+              );
+              wrapper.addEventListener("stop", stopListener);
```
Note: While the assumption isn't obvious, I'd like to note that `onBeforeRedirect` can fire repeatedly for the same request, but `wrapper.addEventListener(..., stopListener)` ensures that each `wrapper` has only one `stopListener`).

willdurand replied:
I'll add a comment

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,233 @@
+              });
+            };
+
+            const listener = ({ requestId, url, redirectUrl }) => {
```
`url` and `redirectUrl` are unused here (but see other comment about tracking the matched URL).

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,199 @@
+  }
+
+  async onRedirectedListener({ addonId, firstUrl, lastUrl }) {
+    if (!firstUrl || !lastUrl) {
```
Any reason for having this check here instead of api.js?

Would it make sense to move this check (and the generation of the public suffix) to the parent? The only thing that can potentially depend on the full URL is the generation of `addonIds`, but that can/should probably also be in the api.js script as mentioned elsewhere (search for "matched URL" in my review).

willdurand replied:
I moved this check to the api code to avoid firing an event. The rest is kept in the background script, which describes the reporting logic.

Rob--W commented on extension/background.js:
```diff
@@ -0,0 +1,199 @@
+
+  getAddonIdsForUrl(url) {
+    for (const pattern of Object.keys(this.matchPatterns)) {
+      const [urlPrefix] = pattern.split("*");
```
Change to `.slice(0, -1)` with comment that `getMatchPatterns` returns the prefix plus `"*"`. Otherwise this logic could break if an extension uses `*` in the URL.

willdurand replied:
if an extension uses `*` in the url, we still get a url prefix, no? (which is what I want)

willdurand committed:
fixes

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,253 @@
+            };
+
+            const listener = ({ requestId, url }) => {
+              const wrapper = ChannelWrapper.getRegisteredChannel(
```
Move inside the if-block

Rob--W commented on extension/api.js:
```diff
@@ -0,0 +1,253 @@
+                Cu.reportError(err);
+              }
+
+              let redirectChain;
```
redirectChain isn't used any more.

willdurand replied:
all of this for nothing 😭

Rob--W reviewed (approved):
r+wc

willdurand committed:
cleanups

willdurand commented:
got a r+ from @mixedpuppy on slack so let's land this now.
